torch/csrc/autograd/autograd_not_implemented_fallback.cpp

```diff
@@ -113,22 +113,31 @@ static void basicAutogradNotImplementedFallbackImpl(
   TORCH_INTERNAL_ASSERT(
       getAutogradFallbackMode() == AutogradFallbackMode::Warn);
 
-  // Keep track of these tensors so we can install a grad_fn that points to them
-  std::vector<const at::Tensor*> tensors_requiring_grad_on_stack;
-  _foreach_tensor(
-      [&](size_t _, size_t idx_arg, const at::Tensor& t) {
-        if (grad_mode && t.requires_grad()) {
-          tensors_requiring_grad_on_stack.push_back(&t);
-        }
-      },
-      stack,
-      stack_start,
-      num_arguments);
-
-  const bool any_input_requires_grad = !tensors_requiring_grad_on_stack.empty();
+  bool any_input_requires_grad = false;
+  if (grad_mode) {
+    _foreach_tensor(
+        [&](size_t _, size_t idx_arg, const at::Tensor& t) {
+          if (t.requires_grad()) {
+            any_input_requires_grad = true;
+          }
+        },
+        stack,
+        stack_start,
+        num_arguments);
+  }
 
   std::shared_ptr<WarnNotImplemented> grad_fn;
   if (any_input_requires_grad) {
+    std::vector<const at::Tensor*> tensors_requiring_grad_on_stack;
+    _foreach_tensor(
+        [&](size_t _, size_t idx_arg, const at::Tensor& t) {
+          if (grad_mode && t.requires_grad()) {
+            tensors_requiring_grad_on_stack.push_back(&t);
+          }
+        },
+        stack,
+        stack_start,
+        num_arguments);
```

Comment:
Can we add a comment that this differs from what VariableType is doing? e.g. in VariableType we gather next edges from ALL tensors even if that tensor does not require grad.
Or, just changing it to be the same as what VariableType might be better (though I don't know what breaks) - and a quick comment in that case would still be good, since this is kind of unexpected, like you could mention the CopySilces thing.

Reply:
We could change it to what VariableType does. This PR is modeled after the existing AutogradNotImpelementedFallback, which only gathers the edges that requires grad.

I tried this out locally before (gather next edges from all tensors), and nothing breaks, but I'll poke at it some more